lib/climate_control.rb

```diff
@@ -1,16 +1,44 @@
-require "climate_control/environment"
 require "climate_control/errors"
-require "climate_control/modifier"
 require "climate_control/version"
+require "monitor"
 
 module ClimateControl
-  @@env = ClimateControl::Environment.new
+  extend self
 
-  def self.modify(environment_overrides = {}, &block)
-    Modifier.new(env, environment_overrides, &block).process
+  SEMAPHORE = Monitor.new
+  private_constant :SEMAPHORE
+
+  def modify(environment_overrides = {}, &block)
+    environment_overrides = environment_overrides.transform_keys(&:to_s)
+
+    SEMAPHORE.synchronize do
+      previous = ENV.to_hash
+      middle = {}
+      copy environment_overrides
+      middle = ENV.to_hash
+      block.call
+    ensure
+      after = ENV
+      (previous.keys | middle.keys | after.keys).each do |key|
+        if previous[key] != after[key] && middle[key] == after[key]
+          ENV[key] = previous[key]
+        end
+      end
+    end
```

Comment:
If the `copy` method were to raise partway through (eg one of the overrides is unassignable), then `ENV` would be modified by the overrides prior to raising, and `middle` would be `{}` as that's what it was initialized to, and it never got to the post override assignment.

So in the `ensure` block, the check `middle[key] == after[key]` will be false for any override that wasn't `nil`, causing the `if` statement to evaluate to `false`, causing the variable to not get restored.

So, I think that `middle` should be initialized to `ENV.to_hash` instead of `{}`.

Reply:
Here is an example:

```ruby
# Run the code from github instead of rubygems
repo_dir = '/Users/josh/ref/climate_control'
Dir.chdir repo_dir
$LOAD_PATH.unshift File.expand_path('lib', repo_dir)
`git log -1 --pretty=oneline`
# => "8b5fab49cb6edb73acafa44bb8491bcd797058ee Add \"About thoughtbot\" section to README\n"

# Example of how setting `middle` to `{}` can cause it to not restore the hash
require 'climate_control'
ENV.merge! 'A' => 'pre a', 'B' => 'pre b'

# Expected behaviour:
ENV.values_at 'A', 'B'  # => ["pre a", "pre b"]
ClimateControl.modify(A: 'overridden a') {}
ENV.values_at 'A', 'B'  # => ["pre a", "pre b"]

# But if setting `B` rasises, then it won't restore "A"
ENV.values_at 'A', 'B'  # => ["pre a", "pre b"]
ClimateControl.modify(A: 'overridden a', B: :not_a_string) {} rescue nil
ENV.values_at 'A', 'B'  # => ["overridden a", "pre b"]
```

Reply:
🤔 actually, I just tried that, and that's not good enough either, b/c if it never gets out of the `copy` command, then `middle` won't contain the overrides.

Reply:
Okay, this worked for me: 
<img width="497" alt="image" src="https://user-images.githubusercontent.com/77495/170794141-b9cfbd98-0503-44c2-b8a5-589ced9c0775.png">

Reply:
Thanks a lot, I will release a new version https://github.com/thoughtbot/climate_control/commit/5e9dd4abce1a8b3fedc77dc144108714f9cbcc0a